File: 2560. 打家劫舍 IV.py
from typing import List
class Solution:
    def minCapability(self, nums: List[int], k: int) -> int:
        def check(x):
            visted = False
            count = 0
            for num in nums:
                if num <= x and not visted:
                    count += 1
                    visted = True
                else:
                    visted = False
            return count >= k
        left,right = min(nums),max(nums)
        while left<right:
            mid = (left+right)//2
            if check(mid):
                right = mid
            else:
                left = mid+1
        return left
    
nums = [2,3,5,9]
k = 2

nums = [2,7,9,3,1] 
k = 2
print(Solution().minCapability(nums,k))

File: 2560. 打家劫舍 IV.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

class Solution {
public:
    int minCapability(vector<int>& nums, int k) {
        int lower = *min_element(nums.begin(), nums.end());
        int upper = *max_element(nums.begin(), nums.end());
        while(lower < upper) {
            int mid = lower + (upper - lower) / 2;
            int count = 0;
            bool visted = false;
            for(int num:nums){
                if(num <= mid && !visted){
                    count++;
                    visted = true;
                }else{
                    visted = false;
                }
            }
            if (count >= k) {
                upper = mid;
            } else {
                lower = mid + 1;
            }
        }
        return lower;
    }
};

int main(){
    Solution s;
    vector<int> nums = {2,7,9,3,1};
    int k = 2;
    int res = s.minCapability(nums,k);
    cout<< res <<endl;
    return 0;
}